In [217]:
import pandas as pd
import numpy as np

# PX

In [218]:
px_db = pd.read_csv("px.csv")

In [219]:
px_db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5512482 entries, 0 to 5512481
Data columns (total 4 columns):
 #   Column      Dtype 
---  ------      ----- 
 0   Unnamed: 0  int64 
 1   pxid        object
 2   age         int64 
 3   gender      object
dtypes: int64(2), object(2)
memory usage: 168.2+ MB


In [220]:
px_db.isnull().any()

Unnamed: 0    False
pxid          False
age           False
gender        False
dtype: bool

In [221]:
px_db['gender'].value_counts()

gender
FEMALE    3095509
MALE      2416973
Name: count, dtype: int64

In [222]:
px_db['gender'].unique()

array(['FEMALE', 'MALE'], dtype=object)

In [223]:
px_db['age'].value_counts()

age
54     150375
29     125362
28     121702
31     120597
32     120355
        ...  
96       2365
97       1870
98       1339
99       1072
100       826
Name: count, Length: 101, dtype: int64

In [224]:
px_db['age'].unique()

array([ 54,   8,   2,  67,  89,  31,  59,  14,  34,  29,  60,  32,  33,
         1,  27,  65,  64,   5,  62,  40,  50,   6,  83,  69,  37,  22,
        42,  78,  55,  48,  81,  66,  58,  82,  44,  43,  30,  79,  38,
        74,  26,  70,  36,   9,  93,  12,  23,  21,  72,   4,  63,  80,
        61,  16,  20,  18,  11,  46,  45,  49,  71,  35,  25,  28,  53,
        57,  41,  52,  86,   3,  24,  47,  90,  10,  87,  77,  51,  73,
        76,  13,  56,  75,  39,  15,   7,  17,  68,  19,   0,  84,  88,
        92,  95,  96,  98,  91,  85,  99,  94,  97, 100], dtype=int64)

In [225]:
# check for duplicates
px_db.duplicated().any()

False

In [226]:
# drop the said duplicates
px_db = px_db.drop_duplicates()

In [227]:
px_db.duplicated().any()

False

In [228]:
# convert all strings to int
px_db['age'] = pd.to_numeric(px_db['age'], errors='coerce').astype('Int64')

In [229]:
# change N/A values to the mode of the dataframe
px_db['age'].fillna(px_db['age'].where(px_db['age'] > 0).mode().iloc[0], inplace=True)

In [230]:
px_db['age'].unique()

<IntegerArray>
[ 54,   8,   2,  67,  89,  31,  59,  14,  34,  29,
 ...
  92,  95,  96,  98,  91,  85,  99,  94,  97, 100]
Length: 101, dtype: Int64

In [231]:
# check for the number of negative values
(px_db['age'] < 0).sum().sum()

0

In [232]:
px_db['age'].isnull().any()

False

In [233]:
# change negative values to the mode of df['age']
px_db['age'] = px_db['age'].mask(px_db['age'] < 0, px_db['age'].mode())

In [234]:
# seeing that there are age more than 100
px_db['age'].unique()

<IntegerArray>
[ 54,   8,   2,  67,  89,  31,  59,  14,  34,  29,
 ...
  92,  95,  96,  98,  91,  85,  99,  94,  97, 100]
Length: 101, dtype: Int64

In [235]:
# replace outliers with mode
px_db['age'] = px_db['age'].mask(px_db['age'] > 100, px_db['age'].mode())

In [236]:
px_db['age'].value_counts()

age
54     150375
29     125362
28     121702
31     120597
32     120355
        ...  
96       2365
97       1870
98       1339
99       1072
100       826
Name: count, Length: 101, dtype: Int64

In [237]:
# drop row that has 'gender' as gender in df
mask = px_db['gender'] == 'gender'

In [238]:
px_db = px_db[~mask]

In [239]:
px_db['gender'].unique()

array(['FEMALE', 'MALE'], dtype=object)

In [240]:
positive_age_mode = px_db['age'].where(px_db['age'] > 0).mode().iloc[0]
px_db['age'].fillna(positive_age_mode, inplace=True)

In [241]:
px_db.shape[0]

5512482

In [242]:
px_db = px_db.drop_duplicates(subset='pxid')

In [243]:
px_db.shape[0]

5512482

In [244]:
px_db.isnull().any()

Unnamed: 0    False
pxid          False
age           False
gender        False
dtype: bool

In [245]:
appoint_db = pd.read_csv("appointments.csv")

In [246]:
new_pxdb = px_db.merge(appoint_db, left_on='pxid', right_on='pxid')[['pxid', 'age', 'gender']]

In [247]:
new_pxdb.shape[0]

320140

In [248]:
new_pxdb

pxid  age  gender
0       519CAC5EFB113B0E451F77E31541A314   36  FEMALE
1       224DEFC6F387F61F307846B09F945EB2   34  FEMALE
2       224DEFC6F387F61F307846B09F945EB2   34  FEMALE
3       224DEFC6F387F61F307846B09F945EB2   34  FEMALE
4       D02F0CFD3B6617FF0FD923C9787D5522   55    MALE
...                                  ...  ...     ...
320135  FED6D4194DA23CA4607410C55D6DF246   66    MALE
320136  018132C688D9CFBAE3DDAF01505243F3   82  FEMALE
320137  018132C688D9CFBAE3DDAF01505243F3   82  FEMALE
320138  018132C688D9CFBAE3DDAF01505243F3   82  FEMALE
320139  018132C688D9CFBAE3DDAF01505243F3   82  FEMALE

[320140 rows x 3 columns]

In [249]:
new_pxdb.duplicated().any()

True

In [250]:
new_pxdb.duplicated().any()

True

In [251]:
new_pxdb.shape[0]

320140

In [252]:
new_pxdb = new_pxdb.drop_duplicates()

In [253]:
new_pxdb.shape[0]

67194

In [254]:
new_pxdb.duplicated().any()

False

In [255]:

# Get the existing pxid in new_pxdb
existing_pxid = new_pxdb['pxid'].unique()

# Select 200,000 random rows from px_db that do not have the same pxid as in new_pxdb
random_rows_px_db = px_db[~px_db['pxid'].isin(existing_pxid)].sample(n=200000, random_state=42, replace=False)

# Concatenate new_pxdb with the randomly selected rows from px_db
new_pxdb = pd.concat([new_pxdb, random_rows_px_db], ignore_index=True)

# Verify the shape of the updated DataFrame
print(new_pxdb.shape[0])

267194


# Clinic

In [30]:
clinics_db = pd.read_csv("clinics.csv", encoding = "ISO-8859-1")

In [31]:
clinics_db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53962 entries, 0 to 53961
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   clinicid      53962 non-null  object
 1   hospitalname  17538 non-null  object
 2   IsHospital    53962 non-null  bool  
 3   City          53962 non-null  object
 4   Province      53962 non-null  object
 5   RegionName    53962 non-null  object
dtypes: bool(1), object(5)
memory usage: 2.1+ MB


In [32]:
clinics_db['hospitalname'].unique()

array(["St. Luke's Medical Center-Global City",
       'Our Lady of Lourdes Hospital', 'Makati Medical Center', ...,
       'Leyte Provincial Hospital',
       "Lopez District Farmer's Hospital, Inc.", 'Alegre Medical Clinic'],
      dtype=object)

In [33]:
clinics_db['City'].unique()

array(['Taguig', 'Manila', 'Makati', 'San Juan', 'Burgos', 'Butuan City',
       'Basco', 'Quezon City', 'Mandaue City', 'Santa Rosa City',
       'San Fernando City', 'Santiago City', 'Batangas City',
       'Iloilo City', 'Botolan', 'Las Piñas', 'Balabac', 'Pasig', 'Bilar',
       'Malabon', 'Caloocan', 'Muntinlupa', 'Candijay', 'San Nicolas',
       'Calape', 'Cebu City', 'Manito', 'Pasay', 'Cabanatuan City',
       'Antipolo City', 'Taytay', 'Cainta', 'Mandaluyong', 'Marikina',
       'San Leonardo', 'Dasmariñas City', 'Bacoor City', 'Malolos City',
       'Parañaque', 'Imus City', 'Silang', 'San Jose del Monte City',
       'San Mateo', 'Davao City', 'Kidapawan City', 'Tagbilaran City',
       'General Trias', 'Balamban', 'Bacolod City', 'Cagayan de Oro',
       'Guagua', 'Lubao', 'Zamboanga City', 'Rizal', 'Santa Cruz',
       'Asuncion', 'Tagum City', 'Cardona', 'Morong', 'Tanay', 'Teresa',
       'Meycauayan City', 'Santo Domingo', 'San Jose City',
       'Trece Martires City',

In [34]:
clinics_db['RegionName'].unique()

array(['National Capital Region (NCR)', 'Ilocos Region (I)',
       'Caraga (XIII)', 'Cagayan Valley (II)', 'Central Visayas (VII)',
       'CALABARZON (IV-A)', 'Western Visayas (VI)', 'Central Luzon (III)',
       'MIMAROPA (IV-B)', 'Bicol Region (V)', 'Davao Region (XI)',
       'SOCCSKSARGEN (Cotabato Region) (XII)', 'Northern Mindanao (X)',
       'Zamboanga Peninsula (IX)', 'Eastern Visayas (VIII)',
       'Cordillera Administrative Region (CAR)',
       'Bangsamoro Autonomous Region in Muslim Mindanao (BARMM)'],
      dtype=object)

In [35]:
clinics_db.duplicated().any()

False

In [36]:
clinics_db['Province'].unique()

array(['Manila', 'Ilocos Sur', 'Agusan del Norte', 'Batanes', 'Cebu',
       'Laguna', 'La Union', 'Isabela', 'Batangas', 'Iloilo', 'Zambales',
       'Palawan', 'Bohol', 'Ilocos Norte', 'Albay', 'Nueva Ecija',
       'Rizal', 'Cavite', 'Bulacan', 'Davao del Sur', 'Cotabato',
       'Negros Occidental', 'Misamis Oriental', 'Pampanga',
       'Zamboanga del Sur', 'Davao del Norte', 'Misamis Occidental',
       'Quezon', 'South Cotabato', 'Negros Oriental', 'Leyte',
       'Camarines Norte', 'Bukidnon', 'Benguet', 'Pangasinan',
       'Camarines Sur', 'Lanao del Norte', 'Oriental Mindoro',
       'Occidental Mindoro', 'Dinagat Islands', 'Davao Occidental',
       'Marinduque', 'Tarlac', 'Angeles', 'Lanao del Sur', 'Samar',
       'Surigao del Norte', 'Agusan del Sur', 'Mountain Province',
       'Nueva Vizcaya', 'Bataan', 'Eastern Samar', 'Aklan', 'Aurora',
       'Cagayan', 'Masbate', 'Kalinga', 'Sorsogon', 'Abra',
       'Davao Oriental', 'Surigao del Sur', 'Zamboanga Sibugay',
       

In [37]:
clinics_db.isnull().any()

clinicid        False
hospitalname     True
IsHospital      False
City            False
Province        False
RegionName      False
dtype: bool

In [38]:
(clinics_db['hospitalname'].isnull()).sum().sum()

36424

In [39]:
# change N/A values to 'Unknown'
clinics_db['hospitalname'].fillna('Unknown', inplace=True)

In [40]:
clinics_db.isnull().any()

clinicid        False
hospitalname    False
IsHospital      False
City            False
Province        False
RegionName      False
dtype: bool

In [41]:
clinics_db.shape[0]

53962

In [42]:
clinics_db = clinics_db.drop_duplicates(subset='clinicid')

In [43]:
clinics_db.shape[0]

53962

After examining the data, it seems that no interpolation technique can be made since it specifies a certain hospital. We can't do dropping of values as well since 36424 rows will be dropped, affecting the data largely. Thus, putting a placeholder, like "Unknown" is the best choice for this dataframe.

# Appointment

In [108]:
appoint_db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9752932 entries, 0 to 9752931
Data columns (total 11 columns):
 #   Column      Dtype 
---  ------      ----- 
 0   pxid        object
 1   clinicid    object
 2   doctorid    object
 3   apptid      object
 4   status      object
 5   TimeQueued  object
 6   QueueDate   object
 7   StartTime   object
 8   EndTime     object
 9   type        object
 10  Virtual     object
dtypes: object(11)
memory usage: 818.5+ MB


In [45]:
appoint_db.head()

pxid                          clinicid  \
0  EF196B348A49FB32DABC9834DC4FAAD9  ADF7EE2DCF142B0E11888E72B43FCB75   
1  EAE3C87D0B33351272F2E9B9B1B56217  1E0F65EB20ACBFB27EE05DDC000B50EC   
2  7C5C93809D626CC702D08F33985B2B58  1E0F65EB20ACBFB27EE05DDC000B50EC   
3  C300C2B9E0E5D4C46E8093BCDBFA05CA  98C39996BF1543E974747A2549B3107C   
4  B3DBE7F9E4DC33CBC5660E0A923CF8E8  77EE3BC58CE560B86C2B59363281E914   

                           doctorid                            apptid  \
0  BB04AF0F7ECAEE4AAE62035497DA1387  C1CC0949B93D00A559F7A0BD38361E80   
1  82AA4B0AF34C2313A562076992E50AA3  6585A31C60A1886FBA1433C50012B504   
2  82AA4B0AF34C2313A562076992E50AA3  7250DCFF615E6580295C7E6ED4322371   
3  AD61AB143223EFBC24C7D2583BE69251  F5BBDCC08E39332F0AC27BB95CF1396A   
4  AD61AB143223EFBC24C7D2583BE69251  55783FB42A02DABC3B2ED239B924DC87   

     status                     TimeQueued            QueueDate  \
0  Complete            2018-04-11 01:59:58  2018-04-10 16:00:00   
1    Queued            2018-04-10 10:34:16  2018-04-08 16:00:00   
2    Queued            2018-04-03 15:26:19  2018-03-30 16:00:00   
3    Queued  2015-11-24 08:01:13.920000000                  NaN   
4    Queued  2015-11-24 08:01:00.420000000                  NaN   

             StartTime              EndTime          type Virtual  
0                  NaN                  NaN  Consultation     NaN  
1  2018-04-09 10:33:00  2018-04-12 05:29:57     Inpatient     NaN  
2  2018-03-31 15:25:00  2018-04-05 15:51:47     Inpatient     NaN  
3                  NaN                  NaN  Consultation     NaN  
4                  NaN                  NaN  Consultation     NaN

In [46]:
appoint_db.status.unique()

array(['Complete', 'Queued', 'NoShow', 'Serving', 'Cancel', 'Skip',
       'Completed', 'Admitted'], dtype=object)

In [47]:
appoint_db['status'].value_counts()

status
Complete     6473675
Queued       2523855
Serving       284774
NoShow        230048
Cancel        213270
Skip           27298
Completed         10
Admitted           2
Name: count, dtype: int64

In [48]:
appoint_db['type'].value_counts()

type
Consultation    9640181
Inpatient        112751
Name: count, dtype: int64

In [49]:
appoint_db['Virtual'].value_counts()

Virtual
False    6276648
True     1523691
Name: count, dtype: int64

In [50]:
appoint_db.isnull().any()

pxid          False
clinicid      False
doctorid      False
apptid        False
status        False
TimeQueued    False
QueueDate      True
StartTime      True
EndTime        True
type          False
Virtual        True
dtype: bool

In [51]:
appoint_db.duplicated().any()

False

In [52]:
# n = 100000
# appoint_db = appoint_db.sample(n, random_state=25)
# appoint_db.head()

In [53]:
appoint_db = appoint_db.merge(px_db, left_on='pxid', right_on='pxid')[['pxid', 'clinicid', 'doctorid', 'apptid', 'status', 'TimeQueued', 'QueueDate', 'StartTime', 'EndTime', 'type', 'Virtual']]


In [54]:
appoint_db.head()

pxid                          clinicid  \
0  FE4A5D5A20EC492D2FC691F126A568AB  3B8D83483189887A2F1A39D690463A8F   
1  FE4A5D5A20EC492D2FC691F126A568AB  BAE3529F0210CC5A603AC708BB5E7797   
2  E6BC8E80685AA4A239617F196F12A0C3  3B8D83483189887A2F1A39D690463A8F   
3  E6BC8E80685AA4A239617F196F12A0C3  3B8D83483189887A2F1A39D690463A8F   
4  E6BC8E80685AA4A239617F196F12A0C3  BAE3529F0210CC5A603AC708BB5E7797   

                           doctorid                            apptid  \
0  ACB3A881C7CE9ABCAE0CE8C99C86A906  04BC9218E072BEEFEBEE9C97B78A35C9   
1  ACB3A881C7CE9ABCAE0CE8C99C86A906  F3BAF3D43E9C6DE6D41850604F4A3A71   
2  ACB3A881C7CE9ABCAE0CE8C99C86A906  F2335AAFB7EE079C7E7306591057C0C9   
3  ACB3A881C7CE9ABCAE0CE8C99C86A906  DA4023B5A8C3F1BD540EA82552F21134   
4  ACB3A881C7CE9ABCAE0CE8C99C86A906  87C3AD392CEF821752AB5191F67D46A7   

    status                     TimeQueued            QueueDate  \
0   Queued  2020-06-22 05:11:38.247000000  2020-07-02 16:00:00   
1  Serving  2020-07-24 06:30:36.910000000  2020-07-24 16:00:00   
2   Queued  2020-06-26 04:51:16.403000000  2020-07-12 16:00:00   
3   Queued  2020-07-13 06:07:13.537000000  2020-07-28 16:00:00   
4  Serving  2020-07-31 02:13:15.870000000  2020-07-31 16:00:00   

                       StartTime                        EndTime          type  \
0  2020-07-02 17:00:38.073000000  2020-07-02 17:15:38.073000000  Consultation   
1  2020-07-25 02:00:36.877000000  2020-07-25 02:15:36.877000000  Consultation   
2  2020-07-12 17:00:15.833000000  2020-07-12 17:15:15.833000000  Consultation   
3  2020-07-29 05:00:13.133000000  2020-07-29 05:15:13.133000000  Consultation   
4  2020-08-01 02:30:56.457000000  2020-08-01 02:45:56.460000000  Consultation   

  Virtual  
0     NaN  
1     NaN  
2     NaN  
3     NaN  
4     NaN

In [55]:
appoint_db.shape[0]

320140

In [56]:
appoint_db=appoint_db.drop_duplicates(subset='apptid')

In [57]:
appoint_db.shape[0]

320140

Appointments seem to have valid data.

# Doctors

In [58]:
doctors_db = pd.read_csv("doctors.csv", encoding = "ISO-8859-1")

In [59]:
doctors_db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60024 entries, 0 to 60023
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   doctorid       60024 non-null  object 
 1   mainspecialty  27055 non-null  object 
 2   age            20028 non-null  float64
dtypes: float64(1), object(2)
memory usage: 1.4+ MB


In [60]:
doctors_db['mainspecialty'].value_counts()

mainspecialty
Internal Medicine                           3812
General Medicine                            2317
Pediatrics                                  1709
Family Medicine                              894
General Physician                            881
                                            ... 
General Physician (Pediatrics and Adult)       1
Marc                                           1
Orthopaedics (Bone and Joint Surgery)          1
Generalists                                    1
Allergist                                      1
Name: count, Length: 3973, dtype: int64

In [61]:
doctors_db['mainspecialty'].unique()


array(['General Medicine', 'Family Medicine', 'Vascular Medicine', ...,
       'Bone alignment therapy', 'Healot Biomekaniks', 'Allergist'],
      dtype=object)

In [62]:
# WIP, also idk how to drop the stuff thats in the list so if yall know how pls lmk ee
doctors_db = doctors_db[~doctors_db["mainspecialty"].isin(["Test", "Endocrinology (test)", "Testing", "test", "abc", "bbbb", "abbb",
 "isiwiwiw", "super saiyan", "Db", "Ll", "Main", "m", "as", "ggg", "SolSee",  "ya", "1", "MainS", "1231", "123", "MS",  "xxx", "asd", "F?s????e?ape?t??",
   "MainSpecialty", "kirurgji", "zxczcxcxzcxz", "ABCD", "b", "L", "RN/A", "123123", "AAAAAAAAAAA", "A", "AAA", "Omomi", "afsf", "Hjj",  "aoeu", "ooooo", 
   "'><img src=x onerror=prompt(1)>", "s","a", "gaga", "man", "M", "Things", "aaaaaaaaaaa", "Com", "nope", "Rrr", "IDK", "B",
   "Aaa", "Yy", "lkjsadlfkj", "kdsjflkfdsjg", "Yg", "Me :)", "Breeding Uroplatus", "Absenos", "qweqwe", "zzz", "nnnnnn", "sdf",
   "Special","n","12", "AGREE", "Dhjdu", "Me", "Md", "Help", "P", "43", "fdgf", "ertryty", "sdaf", "sdfsda",
   "Est qui minim et au", "Hull Engineering", "Ghhhhj", "Computer Science And Technology", "Gsu", "fhhhf",
   "Ttt", "Ship handling", "acctg", "Gh", "Specialty 1",  "Specialty", "Ok", "Human Resources", "Web Development",
   "ABC", "ertertertert ertert", "Degree", "patient", "Test account", ".", "main specialty", "Web Developer",
   "F", "Welder pipe filter", "Patienr", "Sorry wrong sign up", "sdd", "Helping mother to gave birth by normal delivery",
   "Hshshs", "w", "Hahaha", "Test Only", "Babag", "computer science", "accounting", "Test Specilty", "kik", "Sdj",
   "Bdjsp", "finane", "Being Me", "3", "tester", "Major in accountancy", "any", "CS", "I", "Bshshshs",
    "FN", "sccs", "moon light", "No", "User", "asfasfsaf", "awf", "asdf", "Shejsjj", "Elecrical", "dd", "bba", "N", "x","Internet",
    "HACKING","Worker","Her", "f", "Lalala", "Hh", "core", "Fpsedm", "Sales", "cc", "Sample", "Sen", "Level 1 Hospital",
    "Ghetto", "maka", "Sick", "Qui culpa in quo ea", "Bab", "bussiness", "sasa", "yes", "BEEd", "Bsb", "Hhh", "Restaurants",
    "Bvv", "X", "000", "Kakakak", "Fabricator", "English", "Ddsdaa", "222", "Ids", "tests", "Consult", "I need Medicine",
    "Shy caustic  ", "English Language Studies", "dwadawd", "da", "wada", "fawawf", "Housewife", "hba1c", "H", "bjj",
    "hjikjjbh", "Calvin Smith", "Pineapple Processing", "What", "Shaman", "hhhbb", "banking", "Design", "broken colarbone",
    "S", "Hahahahah","Hsha","Hsha", "asdas", "trial", "sir", "asdasd", "jj", "ef", "a/a", "ad", "adasd", "cfvhgbjnk", "sample",
    "sh", "Snsjsj", "das", "bhb", "im not a doctor", "Q", "--", "vv", "Bsbs", "Nil","Ndhdbehbsjshs", "Seafarer", "to", "On",
    "Bbbb", "hah", "ghbb", "need a demtology", "cxxxc", "Mother", "Resting", "2", "medcert", "Aa", "c", "J", "IMMKMLKSMDLKMS",
    "aaa", "Gygghh", "Hshhs", "Xisis", "Not A Doctor", "guycvujb", "hhh", "qqq", "dfghdfjh", "to motivate", "W", "Asksbdkdnd",
    "Shshaha", "k", "nejdhdjdjdh", "For check up", "Check up", "Wrong Entry", "Partner Visa", "fdd", "All", "jsjsjsja", "LLala",
    "Bbb", "Jajan", "Snsnd", "…", "FOLLOW UP", "hdh", "enen", "js", "wqeqwe"
   ])]

doctors_db = doctors_db.replace(to_replace=["docmark_coli@yahoo.com","Leading You To True Health","SURGUYYY","Doctor","logistics", "In", "main","3D", 
"event coordinator", "a/a", "ad", "SPECIALTY", "non", "Loves Kids", "Bed Making","ONLINE KONSULTA","DR.RUSSEL VILLANUEVA",
"FAMED", "Your Health Is Our Priority", "-", "na", "Patient", "Nothing", " ", "Na", "RALPh Roque", "RN", "Jennifer Apurillo Baldillo",
"appointments", "Marc", "N/a", "Trisha Reyes", "Not Sure", "OSH", "Dr. Anna Mitzi Matela", "Vnc","Dra. beverly",
"Valerie Ann  Valdez", "Gonzalez", "NONE", "Wala", "MBA", "Agnes Atas", "Too many to mention", "None ", "Oliver Delgado Castro",
"N/A ", "Valdez", "Faye Garciano", "Need Consultant Of Doctor Before Vaccine", "N.A", "n.a","0", "nan", "VA", "none", "PGI",
"La", "Second opinion", "Luisa marie s cruz", "ERN", "Herrera", "—-", "MeFocal cert", "Nothings ", "Orjto", "Mr.", "Makati med",
"mr", "doc", "Doc", "Cruz", ". ", "Allen Marquez", "131375", "Azarraga", "Skip", "??", "515828", "no", "Nonde", "Fermont Base", 
"Juan Rafael", "Benoza", "Yu", "jana", "Joshua Bacolod", "unknown"
], value="Unknown")

unique_specialty = doctors_db['mainspecialty'].unique()


In [63]:
doctors_db['age'].unique()

array([  41.,   43.,   26.,   34.,   50.,   62.,   38.,   57.,   44.,
         59.,   39.,   49.,   55.,   37.,   nan,   40.,   48.,   51.,
         46.,   45.,   54.,   61.,   60.,   31.,   65.,   36.,   72.,
         64.,   35.,   47.,   33.,   68.,   42.,   73.,   53.,   58.,
         71.,   52.,   56.,   67.,   63.,   66.,   81.,   69.,  160.,
         70., 1048.,   32.,   91.,   75.,   30.,   78.,   76.,   27.,
         29.,   74.,    8.,    7.,   28.,  124.,   80.,  113.,   24.,
         25.,   92.,   77.,  104.,   23.,   83.,    4.,   86.,   88.,
         21.,   22.,   84.,   79.,  944.,   17.,  132.,   85.,  938.])

In [64]:
doctors_db['age'].value_counts()

age
34.0     1203
35.0     1076
33.0      972
37.0      969
36.0      950
         ... 
4.0         1
83.0        1
124.0       1
92.0        1
938.0       1
Name: count, Length: 80, dtype: int64

In [65]:
doctors_db.isnull().any()

doctorid         False
mainspecialty     True
age               True
dtype: bool

In [66]:
doctors_db['mainspecialty'].isnull().sum().sum()

32969

In [67]:
doctors_db['age'] = pd.to_numeric(doctors_db['age'], errors='coerce').astype('Int64')

In [68]:
#replace outliers with mode
doctors_db['age'] = doctors_db['age'].mask(doctors_db['age'] > 100, doctors_db['age'].mode())

In [69]:
doctors_db['age'].unique()

<IntegerArray>
[  41,   43,   26,   34,   50,   62,   38,   57,   44,   59,   39,   49,   55,
   37, <NA>,   40,   48,   51,   46,   45,   54,   61,   60,   31,   65,   36,
   72,   64,   35,   47,   33,   68,   42,   73,   53,   58,   71,   52,   56,
   67,   63,   66,   81,   69,   70,   32,   91,   75,   30,   78,   76,   27,
   29,   74,    8,    7,   28,   80,   24,   25,   92,   77,   23,   83,    4,
   86,   88,   21,   22,   84,   79,   17,   85]
Length: 73, dtype: Int64

In [70]:
# change N/A values to the mode of the dataframe
doctors_db['age'].fillna(value=doctors_db['age'].mode().iloc[0], inplace=True)

In [71]:
doctors_db['age'].isnull().sum().sum()

0

In [72]:
doctors_db.isnull().any()

doctorid         False
mainspecialty     True
age              False
dtype: bool

In [73]:
#replace null values with Unknown string
doctors_db['mainspecialty'].fillna('Unknown', inplace=True)

In [74]:
doctors_db['mainspecialty'].isnull().sum().sum()

0

In [75]:
doctors_db.shape[0]

59602

In [76]:
doctors_db = doctors_db.drop_duplicates(subset='doctorid')

In [77]:
doctors_db.shape[0]

59602

# Exporting the CSV

In [78]:
doctors_db.to_csv('doctors_modified.csv')

In [79]:
px_db.to_csv('px_modified.csv')

In [80]:
clinics_db.to_csv('clinics_modified.csv')

In [81]:
appoint_db.to_csv('appointments_modifed.csv')